vars/withTaskQueue.groovy

```diff
@@ -0,0 +1,149 @@
+import groovy.transform.Field
+
+public static @Field TASK_QUEUES = [:]
+public static @Field TASK_QUEUES_COUNTER = 0
+
+/**
+  withTaskQueue creates a queue of "tasks" (just plain closures to execute), and executes them with your desired level of concurrency.
+  This way, you can define, for example, 40 things that need to execute, then only allow 10 of them to execute at once.
+
+  Each "process" will execute in a separate, unique, empty directory.
+  If you want each process to have a bootstrapped kibana repo, check out kibanaPipeline.withCiTaskQueue
+
+  Using the queue currently requires an agent/worker.
+
+  Usage:
+
+  withTaskQueue(parallel: 10) {
+    task { print "This is a task" }
+
+    // This is the same as calling task() multiple times
+    tasks([ { print "Another task" }, { print "And another task" } ])
+
+    // Tasks can queue up subsequent tasks
+    task {
+      buildThing()
+      task { print "I depend on buildThing()" }
+    }
+  }
+
+  You can also define a setup task that each process should execute one time before executing tasks:
+  withTaskQueue(parallel: 10, setup: { sh "my-setup-scrupt.sh" }) {
+    ...
+  }
+
+*/
+def call(Map options = [:], Closure closure) {
+  def config = [ parallel: 10 ] + options
+  def counter = ++TASK_QUEUES_COUNTER
+
+  // We're basically abusing withEnv() to create a "scope" for all steps inside of a withTaskQueue block
+  // This way, we could have multiple task queue instances in the same pipeline
+  withEnv(["TASK_QUEUE_ID=${counter}"]) {
```

Comment:
Does `withEnv()` handle nested calls correctly? If someone calls `withTaskQueue()` inside of a `withTaskQueue()` the `TASK_QUEUE_ID` will be updated correctly after the inner `withTaskQueue()`?

```
withTaskQueue() {
  print env.TASK_QUEUE_ID // 1
  withTaskQueue() {
	print env.TASK_QUEUE_ID // 2
  }
  print env.TASK_QUEUE_ID // 1
}
```

Reply:
It does, yeah. That's one of the advantages of using `withEnv`, it handles nesting, parallel blocks, etc

```groovy
node('flyweight') {
    withTaskQueue(parallel: 1) {
      print "1 == ${env.TASK_QUEUE_ID}"
      task { print "1 == ${env.TASK_QUEUE_ID}" }
      withTaskQueue(parallel: 1) {
    	print "2 == ${env.TASK_QUEUE_ID}"
    	task { print "2 == ${env.TASK_QUEUE_ID}" }
      }
      "1 == ${env.TASK_QUEUE_ID}"
    }
}
```

gives this output (with irrelevant lines removed)

```
[Pipeline] echo
1 == 1
[Pipeline] echo
2 == 2
[Pipeline] echo
2 == 2
[Pipeline] echo
1 == 1
```